File: ruby/day_3/diagnostic.rb
def find_most_common_bit(bits)
  bits.max_by { |b| bits.count(b) }
end

def find_least_common_bit(bits)
  bits.min_by { |b| bits.count(b) }
end

def calculate_gamma_rate(inputs)
  bits = inputs.map(&:chars)
  (0...bits[0].size).flat_map do |index|
    find_most_common_bit(bits.transpose[index])
  end.join.to_i(2)
end

def calculate_epsilon_rate(inputs)
  bits = inputs.map(&:chars)
  (0...bits[0].size).flat_map do |index|
    find_least_common_bit(bits.transpose[index])
  end.join.to_i(2)
end

def calculate_power_consumption(inputs)
  gamma_rate = calculate_gamma_rate(inputs)
  epsilon_rate = calculate_epsilon_rate(inputs)
  gamma_rate * epsilon_rate
end

File: typescript/src/day_3/diagnostic.ts
export const powerConsumption = (inputs: string[]) : number => {
  const bits: Array<string[]> = inputs.map(input => Array.from(input));
  return gammaRate(bits) * epsilonRate(bits);
}

const gammaRate = (bits: Array<string[]>): number => {
  const bitsArr: string[] = []
  for(let i = 0; i < bits[0].length; i++) {
    const column = bits.map(row => row[i])
    bitsArr.push(mostCommonBit(column))
  }
  return parseInt( bitsArr.join(''), 2 );
}

const epsilonRate = (bits: Array<string[]>): number => {
  const bitsArr: string[] = []
  for(let i = 0; i < bits[0].length; i++) {
    const column = bits.map(row => row[i])
    bitsArr.push(mostCommonBit(column) == '1' ? '0' : '1')
  }
  return parseInt( bitsArr.join(''), 2 );
}

const mostCommonBit = (bits: string[], tiebreaker = '1'): string => {
  const modeMap = new Map();
  let mostCommon = bits[0], maxCount = 1;
  for(let i = 0; i < bits.length; i++) {
    let bit = bits[i];
    let count = modeMap.get(bit);
    modeMap.set(bit, count ? count + 1 : 1)

    count = modeMap.get(bit);
    if(count > maxCount || (count == maxCount && bit == tiebreaker)) {
        mostCommon = bit;
        maxCount = modeMap.get(bit);
    }
  }
  return mostCommon;
}
